src/nwb/base/TimeSeries.hpp

```diff
@@ -65,79 +44,79 @@ class TimeSeries : public Container
+   *                   the specified ‘unit’
+   * @param resolution Smallest meaningful difference between values in data,
+   *                   stored in the specified by unit
+   * @param offset Scalar to add to the data after scaling by ‘conversion’ to
+   *               finalize its coercion to the specified ‘unit'
    */
-  void initialize();
+  void initialize(const IO::BaseDataType& dataType,
+                  const std::string& unit,
+                  const std::string& description = "no description",
+                  const std::string& comments = "no comments",
+                  const SizeArray& dsetSize = SizeArray {0},
+                  const SizeArray& chunkSize = SizeArray {1},
+                  const float& conversion = 1.0f,
+                  const float& resolution = -1.0f,
+                  const float& offset = 0.0f);
 
   /**
    * @brief Pointer to data values.
    */
-  std::unique_ptr<BaseRecordingData> data;
+  std::unique_ptr<IO::BaseRecordingData> data;
 
   /**
    * @brief Pointer to timestamp values.
    */
-  std::unique_ptr<BaseRecordingData> timestamps;
+  std::unique_ptr<IO::BaseRecordingData> timestamps;
 
   /**
    * @brief Data type of the data.
    */
-  BaseDataType dataType;
+  IO::BaseDataType dataType;
 
   /**
    * @brief Data type of the timestamps (float64).
    */
-  BaseDataType timestampsType = BaseDataType::F64;
-
-  /**
-   * @brief Base unit of measurement for working with the data. Actual stored
-   * values are not necessarily stored in these units. To access the data in
-   * these units, multiply ‘data’ by ‘conversion’ and add ‘offset’.
-   */
-  std::string unit;
-
-  /**
-   * @brief The description of the TimeSeries.
-   */
-  std::string description;
-
-  /**
-   * @brief Human-readable comments about the TimeSeries.
-   */
-  std::string comments;
-
-  /**
-   * @brief Size used in dataset creation. Can be expanded when writing if
-   * needed.
-   */
-  SizeArray dsetSize;
-
-  /**
-   * @brief Chunking size used in dataset creation.
-   */
-  SizeArray chunkSize;
-
-  /**
-   * @brief Scalar to multiply each element in data to convert it to the
-   * specified ‘unit’.
-   */
-  float conversion;
-
-  /**
-   * @brief Smallest meaningful difference between values in data, stored in the
-   * specified by unit.
-   */
-  float resolution;
-
-  /**
-   * @brief Scalar to add to the data after scaling by ‘conversion’ to finalize
-   * its coercion to the specified ‘unit’.
-   */
-  float offset;
-
-  /**
-   * @brief The starting time of the TimeSeries.
-   */
-  float startingTime = 0.0;
+  IO::BaseDataType timestampsType = IO::BaseDataType::F64;
+
+  inline std::string dataPath() const
+  {
+    return (m_path + std::string("/data"));
+  }
+
+  template<typename VTYPE = std::any>
+  inline std::unique_ptr<
+      IO::ReadDataWrapper<AQNWB::Types::StorageObjectType::Dataset, VTYPE>>
+  dataLazy() const
```

Comment:
In #91 I replaced `dataLazy` (and other read fields) with the `DEFINE_FIELD` macro to create these access methods. I changed the name of the functions there to use the naming convention `read<FieldPath>`, e.g., `readData` or `readTimestamps` . To avoid confusion for attributes I used the full path in naming the field, e.g., `readDataUnit` and `readTimestampsUnit`.

I don't think we want to store the read wrappers on the container. They are lightweight and a user may need to use a data type different than the default, so creating the read wrapper on request is useful. I.e., I don't think it is strictly necessary for `BaseRecordingData` and `ReadDataWrapper` to have a common base. We could rename `TimeSeries.data` to `TimeSeries.recordData` to make them more consistent, or we could leave it as is and say that attributes without a prefix are for acquisition and with the prefix are for read only.

With all of this said, I think it is worth discussing whether having a common base class for BaseRecordingData and ReadDataWrapper would be useful.